### ISSUE TITLE
beta.16 TypeError: AsyncTestZoneSpec is not a constructor

### ISSUE THREAD

**Issue by micaelmbagira** [angular/angular#8232]
When I run a test, I have 
> 	TypeError: AsyncTestZoneSpec is not a constructor
	    at runInAsyncTestZone (/sandbox/webapp2/jspm_packages/npm/angular2@2.0.0-beta.16/src/testing/testing.js: 100:24)
	    at Object.eval (/sandbox/webapp2/jspm_packages/npm/angular2@2.0.0-beta.16/src/testing/testing.js:113:17)

because of https://github.com/angular/angular/blob/master/modules/angular2/src/testing/testing.ts#L128-L129

**Comment by KillerCodeMonkey** [comment]
Take a look in the changelog.

There is no injectAsync method in beta 16.

You should use `async` and `inject` instead.

And you have to use `fakeAsync` as test-wrapper if you need an own testZone.

https://github.com/angular/angular/blob/master/CHANGELOG.md#breaking-changes

**Comment by micaelmbagira** [comment]
I know and I updated my test but still. Here is my test:
````
import {Hello} from './hello.js';
import {describe, it, expect, async, inject, TestComponentBuilder} from 'angular2/testing';

describe('hello component', () => {
  it('should render \'Hello\'', async(inject([TestComponentBuilder], tcb => {
    tcb.createAsync(Hello)
      .then(fixture => {
        fixture.detectChanges();
        const footer = fixture.nativeElement;
        expect(hello.querySelector('a').textContent.trim()).toBe('Hello');
      });
  })));
});
````

**Comment by KillerCodeMonkey** [comment]
tcb.createAsync creates already an async context --> you do not need to wrap your test in "async"

**Comment by micaelmbagira** [comment]
Right, thanks !

**Comment by pkozlowski-opensource** [comment]
Hmm, I'm sure I've got something wrong on my end but now I see tests passing when those should be failing... For example this one is green:

```
it('should fail but it does not', inject([TestComponentBuilder], (tcb: TestComponentBuilder) => {
         const html = '<div></div>';

        return tcb.overrideTemplate(TestComponent, html).createAsync(TestComponent).then((fixture) => {
          expect(true).toBeFalsy();
         });
       }));
```

What gives?

**Comment by choeller** [comment]
Yeah, same here... I already tried trillions of combinations, but either I get `AsyncTestZoneSpec is not a function` or all my tests are green, even if they shouldn't be...

**Comment by micaelmbagira** [comment]
I think the `return` is no longer needed.

**Comment by KillerCodeMonkey** [comment]
Yep, i tested it deeper and it tcb.createAsync seems to wrap all stuff in Async call, but simple "inject()" wants to run synchronously. It is not waiting...

**Comment by pkozlowski-opensource** [comment]
Right, but `return` or not I can't make the simplest test fail. Still digging to see if this is a timing issue ('not waiting') or sth else

**Comment by micaelmbagira** [comment]
Right, same for me.

**Comment by micaelmbagira** [comment]
I think `async` should stay because the changelog says so. So maybe the problem is the inital one : `TypeError: AsyncTestZoneSpec is not a constructor` or `TypeError: _FakeAsyncTestZoneSpecType is not a constructor` (if you use `fakeAsync` instead)

**Comment by tleguijt** [comment]
I get the same ```TypeError: AsyncTestZoneSpec is not a constructor```.
All my ```injectAsync(..)``` calls are replaced with ```async(inject(...))```

**Comment by pkozlowski-opensource** [comment]
OK, I see what is going on. Seems like `AsyncTestZoneSpec` is created as part of https://github.com/angular/zone.js/blob/master/dist/async-test.js so you need to include it as part of your Karma setup as well. 

I will shortly push a complete example to https://github.com/ng-bootstrap/core and open a PR with breaking changes update.

**Comment by choeller** [comment]
@pkozlowski-opensource shouldn't this be part of angular2-polyfills.js? I don't include any files from zone directly...

**Comment by micaelmbagira** [comment]
Well I tried to add `node_modules/zone.js/dist/zone.js` and `node_modules/zone.js/dist/async-test.js` to karma files but still same error.

**Comment by ericmdantas** [comment]
@micaelmbagira I did that and now I got:

```
  Uncaught TypeError: Cannot read property 'wrap' of undefined
  at C:/Users/path-to-module/node_modules/angular2/bundles/angular2-polyfills.js:744
```

**Comment by ericmartinezr** [comment]
@ericmdantas you have twice zone (zone.js and angular2-polyfills), I would remove angular2-polyfills and let only zone (and it should be above everything).

**Comment by ericmdantas** [comment]
@ericmartinezr, no luck. If I remove `angular-polyfills` and let `zone.js` and `async-test`, I get:

```
Missing error handler on `socket`.
TypeError: (msg || "").replace is not a function[...]
```

**Comment by choeller** [comment]
So I got it finally working on my side. What I do right now is:

```js
      { pattern: 'node_modules/angular2/bundles/angular2-polyfills.js', included: true, watched: true },
      { pattern: 'node_modules/zone.js/dist/async-test.js', included: true, watched: true },
```
in karma.conf

and 

```ts
  iit('should fail', async(inject([TestComponentBuilder],
      (_tcb:TestComponentBuilder) => {
        return _tcb.createAsync(EditTaskComponent).then(fixture =>  {
            expect(1).toBe(2)
          });
      }))
  );
```
in the test. This now correctly fails with ` Expected 1 to be 2.`

**Comment by pkozlowski-opensource** [comment]
@choeller good, this is the setup that worked for me as well. See the full diff / setup in https://github.com/ng-bootstrap/core/pull/183

Info about Karma setup needs to definitively be added to changelog, will open a PR.

**Comment by choeller** [comment]
@pkozlowski-opensource When doing real async tests I also had to add: fake-async-test:
```
      { pattern: 'node_modules/angular2/bundles/angular2-polyfills.js', included: true, watched: true },
      { pattern: 'node_modules/zone.js/dist/async-test.js', included: true, watched: true },
      { pattern: 'node_modules/zone.js/dist/fake-async-test.js', included: true, watched: true },
```
otherwise I got: `_FakeAsyncTestZoneSpecType`

**Comment by schippie** [comment]
@choeller Indeed I also put it https://github.com/angular/angular/pull/7735#issuecomment-214720413 as i ran into it as well and managed to figure it out myself. But this should really have been part of the changelog.

**Comment by andrashatvani** [comment]
Even after making the suggested changes I still see a few test errors, among others: "Error: Cannot make XHRs from within a fake async test."

**Comment by ericmdantas** [comment]
@andrashatvani how's the `files` session in your karma file?

**Comment by andrashatvani** [comment]
@ericmdantas I use browserify and therefore it's pretty short:
files: [
            // Traceur is needed for PhantomJS
            'node_modules/traceur/bin/traceur-runtime.js',
            'node_modules/angular2/bundles/angular2-polyfills.js',
            'node_modules/zone.js/dist/async-test.js',
            'node_modules/zone.js/dist/fake-async-test.js',
            'app/test/test-setup.js',
            'app/**/*.spec.js',
            'app/**/*.html'
        ]

**Comment by choeller** [comment]
also `tick()` seems to be broken within new zones. I already added another issue for it: https://github.com/angular/angular/issues/8251

**Comment by ericmdantas** [comment]
@andrashatvani I'd suggest you didn't have `**/*.html` but it's there, so I don't know.

**Comment by pkozlowski-opensource** [comment]
OK, changelog was updated with the proper instructions for karma setup via #8253. Closing this one.

**Comment by cdcooksey** [comment]
This issue persists.  The changelog doesn't work. 

`Error: Cannot find module "node_modules/zone.js/dist/async-test.js"`

I verified that file is indeed in `node_modules/zone.js/dist/`, but it will not load.  I put it in my `karma.config.js` `files: []` array, and it failed in this way.  I added it to the top of my test as 

```typescript
import 'node_modules/zone.js/dist/async-test.js'
```

Still fails with `Error: Cannot find module "node_modules/zone.js/dist/async-test.js"`

The ChangeLog also does not demonstrate how to use this for a directive's integration test.
